docs/design_docs/proxy.md

```diff
@@ -0,0 +1,511 @@
+
+![task_scheduler_2](./graphs/task_scheduler_2.png)
+
+The following figure is a schematic diagram of taskScheduer's scheduling of DqQueue.
+
+![task_scheduler_1](./graphs/task_scheduler_1.png)
+
+The tasks in DqQueue can be scheduled in parallel. In a scheduling process, taskScheduler will execute several tasks
+concurrently.
+
+In order to facilitate the channelsTimeTicker component to obtain the time point information corresponding to all
+DmChannels, the taskScheduer needs to maintain a copy of the time statistics of the physical channels of all currently
+unexecuted and executing tasks in the DmQueue. The member pChanSatisticsInfos is a map containing the mapping from pChan
+to pChanStatInfo pointers. Among them, pChan is an alias of string, and pChanStatInfo is a custom structure, defined as
+follows:
+
+```go
+type Timestamp = uint64
+type pChan = string
+
+type pChanStatInfo struct {
+   maxTs Timestamp
+   minTs Timestamp
+  tsSet map[Timestamp] struct{}
+}
+```
+
+pChan represents the name of the physical channel corresponding to the DmChannel. The pChanStatInfo structure contains 3
+members minTs, maxTs and tsSet. minTs and maxTs respectively represent the minimum and maximum timestamps of all tasks
+related to the pChan in the current DmQueue. tsSet represents the set of timestamps of all tasks in DmQueue. When the
+task enters the DmQueue, it will call the addPChanStats method of the queue, add the task's own timestamp to the
+collection of pChanStatInfo's tsSet, and use the task's own maxTs to recalculate the pChanStatInfo's maxTs. The
+calculation method is if its own maxTs is greater than pChanStatInfo MaxTs in pChanStatInfo, then update maxTs in
+pChanStatInfo. Since the newly added timestamp is definitely greater than the existing timestamp, there is no need to
+update minTs. When the PopActivateTask interface of the Queue is called to take the task out of the DmTaskQueue, the
+popPChanStats interface of the Queue is called to delete the task timestamp from the timestamp set tsSet of
+pChanStatInfo, and recalculate the minimum timestamp minTs.
+
+DmQueue maintains the timestamp statistics of pachas and provides the method getPChanStatsInfo to the caller.
+pChanStatistics is defined as follows:
+
+```go
+type pChanStatistics struct {
+	minTs Timestamp
+	maxTs Timestamp
+}
+
+func (queue *DmTaskQueue) getPChanStatsInfo() (map[pChan]*pChanStatistics, error)
+```
+
+The return value of this method includes a pChan to pChanStatistics pointer mapping. pChanStatistics includes two
+members minTs and maxTs, which respectively represent the minimum and maximum timestamps of all tasks that have not been
+completed on pChan. The channelsTimeTicker collects timestamp information mainly depends on this method.he coroutine,
+the awakened coroutine reduces the result and sends the final result back to the client.
+
+taskScheduler is also responsible for collecting the results of search requests. For each search request, when the proxy
```

Comment:
Is there a limit about how many goroutines in total?

Reply:
now the limit is a hardcode number: 1024